### PR TITLE
Fix OOM error and add support for JPEG-2000 compressed image data

### PR DESCRIPTION
**Pull request by t1ld3x**
Changes:
- Fix OOM error and add support for JPEG-2000 compressed image data
- Add unit tests

### CODE REVIEW

File: src/lib/openjp2/helper.c [hunk 1]
```diff
@@ -206,5 +278,100 @@ opj_image_t* decode_j2k(
...
+    {
+        return 0;
+    }
+    // https://github.com/malaterre/GDCM/blob/master/Source/MediaStorageAndFileFormat/gdcmJPEG2000Codec.cxx#L656
+    const char jp2magic[] = "\x00\x00\x00\x0C\x6A\x50\x20\x20\x0D\x0A\x87\x0A";
+    if( memcmp( src, jp2magic, sizeof(jp2magic) ) == 0 )
+    {
+        /* JPEG-2000 compressed image data ... sigh */
+        fprintf(stderr, "J2K start like JPEG-2000 compressed image data instead of codestream" );
+        decod_format = JP2_CFMT;
```
Comment (t1ld3x) [inline src/lib/openjp2/helper.c:305]: add support for JPEG-2000 compressed image data

File: src/lib/openjp2/helper.c [hunk 2]
```diff
@@ -197,6 +268,7 @@ opj_image_t* opj_decompress(
     /* free remaining structures */
     if (l_codec) {
         opj_destroy_codec(l_codec);
+        opj_stream_destroy(l_stream);
```
Comment (t1ld3x) [inline src/lib/openjp2/helper.c:271]: free memory allocated for j2k byte stream

File: gojp2/helper.go [hunk 3]
```diff
@@ -62,6 +62,7 @@ func DecodeJ2KImage(data []byte) (int, int, [][]int, error) {
 		return 0, 0, nil, fmt.Errorf("data is empty")
 	}
 	var opjImage uintptr = Xdecode_j2k(tls, uintptr(unsafe.Pointer(&data[0])), uint64(len(data)))
+	defer Xopj_image_destroy(tls, opjImage)
```
Comment (t1ld3x) [inline gojp2/helper.go:65]: free memory allocated for j2k byte image

File: gojp2/j2k_test.go [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+	in                   string
+	out                  string
+	w, h, prec, numcomps int32
+}
+
+func TestMonoDecompresing(t *testing.T) {
+	for _, test := range []testify{
+		{
+			in:       "a1_mono.j2c",
+			out:      "a1_mono.ppm",
```
Comment (favadi) [inline gojp2/j2k_test.go]: those files should be moved to testdata.

File: src/lib/openjp2/helper.c [hunk 5]
```diff
@@ -136,11 +197,21 @@ opj_image_t* opj_decompress(
...
+    {
+    case J2K_CFMT:
+        l_codec = opj_create_decompress(OPJ_CODEC_J2K);
+        break;
+    case JP2_CFMT:
+        l_codec = opj_create_decompress(OPJ_CODEC_JP2);
+        break;
+    default:
+        fprintf(stderr, "Impossible happen" );
+        return 0;
```
Comment (favadi) [inline src/lib/openjp2/helper.c:211]: should we return non-0 in this case?

**Review by favadi — APPROVED** [review]
I left some minor comments.

File: src/lib/openjp2/helper.c (hunk 5 above)
Comment (t1ld3x) [reply src/lib/openjp2/helper.c:211]: In this context, 0 means FALSE not an exit code.

File: gojp2/j2k_test.go (hunk 4 above)
Comment (t1ld3x) [reply gojp2/j2k_test.go]: fixed in bf57199

File: src/lib/openjp2/helper.c (hunk 5 above)
Comment (favadi) [reply src/lib/openjp2/helper.c:211]: Oops, ignore it then.

**Pull request merged**